subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c

```diff
@@ -124,47 +124,60 @@ static void print_location(struct nrf_modem_gnss_agps_data_location *data)
 	printk("\tconfidence: %d\n", data->confidence);
 }
 
-static void print_integrity(struct nrf_modem_gnss_agps_data_integrity *data)
+static void print_gps_integrity(struct nrf_modem_gnss_agps_data_integrity *data)
 {
 	printk("integrity\n");
-	printk("\tintegrity_mask: %d\n", data->integrity_mask);
+	printk("\tintegrity_mask: 0x%08x\n", data->integrity_mask);
+}
+
+static void print_gnss_integrity(struct nrf_modem_gnss_agnss_data_integrity *data)
+{
+	printk("integrity\n");
+	for (int i = 0; i < data->signal_count; i++) {
+		printk("\tsignal_id: %d, integrity_mask: 0x%16llx\n",
+		       data->signal[i].signal_id, data->signal[i].integrity_mask);
+	}
 }
 
 void agps_print(uint16_t type, void *data)
 {
 
 	switch (type) {
-	case NRF_MODEM_GNSS_AGPS_UTC_PARAMETERS: {
-		print_utc((struct nrf_modem_gnss_agps_data_utc *)data);
+	case NRF_MODEM_GNSS_AGNSS_GPS_UTC_PARAMETERS: {
+		print_utc((struct nrf_modem_gnss_agnss_gps_data_utc *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_EPHEMERIDES: {
-		print_ephemeris((struct nrf_modem_gnss_agps_data_ephemeris *)data);
+	case NRF_MODEM_GNSS_AGNSS_GPS_EPHEMERIDES: {
+		print_ephemeris((struct nrf_modem_gnss_agnss_gps_data_ephemeris *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_ALMANAC: {
-		print_almanac((struct nrf_modem_gnss_agps_data_almanac *)data);
+	case NRF_MODEM_GNSS_AGNSS_GPS_ALMANAC: {
+		print_almanac((struct nrf_modem_gnss_agnss_gps_data_almanac *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_KLOBUCHAR_IONOSPHERIC_CORRECTION: {
-		print_klobuchar((struct nrf_modem_gnss_agps_data_klobuchar *)data);
+	case NRF_MODEM_GNSS_AGNSS_KLOBUCHAR_IONOSPHERIC_CORRECTION: {
+		print_klobuchar((struct nrf_modem_gnss_agnss_data_klobuchar *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_NEQUICK_IONOSPHERIC_CORRECTION: {
-		print_nequick((struct nrf_modem_gnss_agps_data_nequick *)data);
+	case NRF_MODEM_GNSS_AGNSS_NEQUICK_IONOSPHERIC_CORRECTION: {
+		print_nequick((struct nrf_modem_gnss_agnss_data_nequick *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS: {
+	case NRF_MODEM_GNSS_AGNSS_GPS_SYSTEM_CLOCK_AND_TOWS: {
 		print_clock_and_tows(
-			(struct nrf_modem_gnss_agps_data_system_time_and_sv_tow *)data);
+			(struct nrf_modem_gnss_agnss_gps_data_system_time_and_sv_tow *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_LOCATION: {
-		print_location((struct nrf_modem_gnss_agps_data_location *)data);
+	case NRF_MODEM_GNSS_AGNSS_LOCATION: {
+		print_location((struct nrf_modem_gnss_agnss_data_location *)data);
 		break;
 	}
 	case NRF_MODEM_GNSS_AGPS_INTEGRITY: {
```

Comment:
`NRF_MODEM_GNSS_AGPS_LOCATION` seem a bit misplaced here now, maybe(?): 
```suggestion
	case NRF_MODEM_GNSS_AGNSS_INTEGRITY_GPS: {
```

Reply:
The GNSS API currently contains two types for satellite integrity data, `NRF_MODEM_GNSS_AGPS_INTEGRITY` and `NRF_MODEM_GNSS_AGNSS_INTEGRITY`. `NRF_MODEM_GNSS_AGPS_INTEGRITY` contains data only for GPS satellites while `NRF_MODEM_GNSS_AGNSS_INTEGRITY` can contain data for multiple constellations, but it's only supported from MFW v2.0.0 onwards.

These could be combined in the API, but then the modemlib implementation needs to decide which data type to use in the GNSS RPC API. The decision could be done based on whether integrity is given for only GPS satellites or not, or the implementation could try first using the newer data type and fallback to the older data type if the first request is rejected by the modem.

The application knows from the A-GNSS data request if modem supports assistance for other systems than GPS. It can then request correct assistance data from the network and write integrity assistance using the correct type. This is why I decided to keep two types for integrity data in the GNSS API. That can still be changed, though. What do you think, @eivindj-nordic  and @trantanen?

Reply:
Hmm, good question. 

TL;DR
If there is only one request type (NRF_MODEM_GNSS_AGNSS_INTEGRITY_REQUEST) I think it should be enough with one data type (NRF_MODEM_GNSS_AGNSS_INTEGRITY).

From what I can see, we only have `NRF_MODEM_GNSS_AGNSS_INTEGRITY_REQUEST` that is common for both(?), but we have two types to write. I think it should be ok to try `RPC_GNSS_AGNSS_REQ_DATA_INTEGRITY` first, and if that is not successful try with `RPC_GNSS_AGNSS_REQ_DATA_INTEGRITY_V1`. If we store the result the firs time, we only have to try and fail once. On the other hand I guess this is not a command that is called often, so one extra RPC request now and then should not create much extra load.  

FYI we are looking at the option/need to have separate libmodem binaries for nRF9160 and nRF91x1. Nothing is decided here yet, and timeline is uncertain, but if we choose to do so, then we can select the correct data type dependent on what target it is built for and won't have to try and fail on nRF9160.

Reply:
@eivindj-nordic I tried changing this so, that there would be only one type for integrity data, but realized that this is a bit problematic. The SUPL library gives the data in the old format, so essentially changing the structure of data type 8 (`NRF_MODEM_GNSS_AGPS_INTEGRITY`) causes problems. This would either require a new release of the SUPL library (which is released separately as a binary delivery) or conversion in all applications using the library. Currently, the handler on the application side just passes the data unmodified to `nrf_modem_gnss_agps_write()`:

https://github.com/nrfconnect/sdk-nrf/blob/64ebb31f39903a83a3d611cff1ae332d30c109f1/samples/cellular/modem_shell/src/gnss/gnss.c#L790

But to make it work with the new data structure, the handler would have to convert integrity data from `struct nrf_modem_gnss_agps_data_integrity` to `struct nrf_modem_gnss_agnss_data_integrity` before calling `nrf_modem_gnss_agps_write()`.

I'm leaning towards keeping both types in the API...

Reply:
I see...
I'm not familiar with the SUPL library. Do you know who is maintaining the SUPL library, and how much work it would entail releasing a new version of it? 

It is definitely easier keeping both in that sense. On the other hand I think it would be cleaner to only have one type for integrity data in the API. I don't have a strong opinion of it, perhaps that can be cleaned up in a separate PR later, after we split the binaries for nRF9160 and nRF91x1(?).

Reply:
I am. 😄 We've released a new version when there has been significant fixes or improvements for the library. Because it's released as a separate binary delivery (see https://www.nordicsemi.com/Products/Development-hardware/nRF9160-DK/Download?lang=en#infotabs), there is some extra work involved.

From API usage perspective this is pretty straight forward, I think. `NRF_MODEM_GNSS_AGPS_INTEGRITY` can be used to only write GPS integrity, while `NRF_MODEM_GNSS_AGNSS_INTEGRITY` can be used to write integrity for all supported signals. Normally users don't have to even use these directly, but it's handled by the nRF Cloud library. If you have a look at the implementation in `agps_send_to_modem()` in `nrf_cloud_agps.c`, there's really no extra logic required. GPS integrity is always written using `NRF_MODEM_GNSS_AGPS_INTEGRITY` and it works with both v1.x.x and v2.x.x MFWs. QZSS integrity is always written with `NRF_MODEM_GNSS_AGNSS_INTEGRITY`. It doesn't work with MFW v1.x.x, but it does not support QZSS assistance and does not ask for it either.

I do agree that it would be cleaner to have only one type, but because of the SUPL library dependency I would rather not break backwards compatibility with `NRF_MODEM_GNSS_AGPS_INTEGRITY`.